README.md
# rpctools

The rpctools package provides RPC clients (currently JSON-RPC, eventually XML-RPC too) with enhancements such
as improved SSL support and connection pooling.

Currently the only protocol implemented is JSON-RPC.  The enhanced SSL support is simply that these libraries
can present client certificates for autentication and can be setup to require a trusted SSL connection with 
the server (validating CA and hsotname matches).

This project was created to address the need for better SSL support when using JSON-RPC, in particular to allow for 
the use of client certificates in authentication and to validate server certificates.  Python's SSL defaults tend
to forgo security for the sake of "ease of use"; we actually want to use SSL for its security features.

** IMPORTANT **
This library is alpha-quality and should be considered a preview.  It needs unit/functional tests (and will be getting
them).  While there are no plans to change the API for JSON-RPC, but more features will be added and backwards-incompatible 
changes may be introduced. 

## Known Limitations / Plans

- Unit/functional tests are current #1 priority.
- Currently only JSON-RPC is supported.  We plan to add support for XML-RPC also.
- The connection pooling system should be considered alpha-quality.  We would love feedback, but don't expect 
  a bug-free experience.
- We plan to support Python 3.x (no testing has been done, but using 2to3 may work with little modification)

## Installation

Via distribute/setuptools:

    easy_install rpctools

Or more traditionally:

    python setup.py install
    
## Basic Usage

The JSON-RPC API is modeled after Python's [xmlrpclib](http://docs.python.org/2/library/xmlrpclib.html) API and
also draws some inspiration from other Python clients -- e.g. the [Redis client](https://github.com/andymccurdy/redis-py)
for connection pooling.

### Connecting to an HTTPS Server
	
```python
from rpctools.jsonrpc import ServerProxy, Fault

proxy = ServerProxy(uri='https://example.com/jsonrpc',
                    ca_certs='/path/to/ca-bundle.crt', # PEM-encoded contatenated set of CA certificates
                    validate_cert_hostname=True # (This is also the default.)
                    )

try:
    proxy.someServerMethod(param1, param2)
except Fault:
    # Fault instances are used to communicate server-side exceptions. 
    raise
```

### ... with basic auth

The underlying httplib library supports providing basic auth in the URI:

```python
from rpctools.jsonrpc import ServerProxy, Fault

proxy = ServerProxy(uri='https://foo:pass@example.com/jsonrpc',
                    ca_certs='/path/to/ca-bundle.crt'
                    )

try:
    proxy.requiresAuth(param1, param2)
except Fault:
    # Fault instances are used to communicate server-side exceptions. 
    raise
``` 

### ... or client certs

```python
from rpctools.jsonrpc import ServerProxy, Fault

proxy = ServerProxy(uri='https://example.com/jsonrpc',
                    key_file='/path/to/client.key', # PEM-encoded
                    cert_file='/path/to/client.crt', # PEM-encoded
                    ca_certs='/path/to/ca-bundle.crt'
                    )

try:
    proxy.someServerMethod(param1, param2)
except Fault:
    # Fault instances are used to communicate server-side exceptions. 
    raise
``` 

### ... connection pooling (ALPHA!)

If you are going to make many repeated calls to the server, you may find it helpful
to use the connection pool feature. 

```python
from rpctools.jsonrpc import ServerProxy, Fault

proxy = ServerProxy(uri='http://example.com/jsonrpc',
                    pool_connections=True)

for (param1, param2) in some_params_list:
	try:
	    proxy.someServerMethod(param1, param2)
	except Fault:
	    # Fault instances are used to communicate server-side exceptions. 
	    raise
``` 


docs/news.rst
News: jsonrpcclient
===================

.. contents::

0.1.0
-----
* Initial alpha-quality release.

rpctools/jsonrpc/exc.py
"""
Exception class hierarchy.
"""

__license__ = """Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""


class JsonRpcError(Exception):
    """
    Base class for *errors* (not faults) in jsonrpclib.
    
    "Errors" are distinguished by "faults" in that errors are problems in communicating
    with the server, whereas faults are problems reported back from the remote server.  
    """

class ConnectionError(JsonRpcError):
    """
    Indicates an error at the TCP connection level.
    
    This exception is raised when a socket.error is raised from the underlying httplib layer.
    """
    
class ProtocolError(JsonRpcError):
    """
    Indicates an HTTP protocol error.
    
    This exception is raised when a non-200 response is received from the server.
    """
    def __init__(self, url, errcode, errmsg, headers):
        JsonRpcError.__init__(self, url, errcode, errmsg, headers)
        self.url = url
        self.errcode = errcode
        self.errmsg = errmsg
        self.headers = headers
    def __repr__(self):
        return (
            "<ProtocolError for %s: %s %s>" %
            (self.url, self.errcode, self.errmsg)
            )
        
class ResponseError(JsonRpcError):
    """
    Indicates an invalid JSON response format.
    
    The JSON-RPC spec clearly describes how responses should be structured.  This 
    exception will be raised if the response does not match this structure.
    """
    pass

class Fault(Exception):
    """
    Represents an error that happened on the remote application that is being passed
    back over the RPC channel.
    
    For example, if I request a book with ISBN 12345 from a library API, but the library
    raises an Exception that the book does not exist, that will be represented as a 
    Fault in the consumer code. 
    """
    def __init__(self, errcode, errmsg, **extra):
        super(Fault, self).__init__(errmsg
                                    )
        self.errcode = errcode
        self.errmsg = errmsg
    
    def __repr__(self):
        return ("<%s %s: %r>" % (self.__class__.__name__, self.errcode, self.errmsg))

rpctools/jsonrpc/pool.py
"""
Support for connection pooling.  (BETA!)
"""
from __future__ import absolute_import
from threading import local as ThreadLocal

__license__ = """Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""

class Pool(ThreadLocal):
    """
    A thread-local subclass that is responsible for managing a pool of connections,
    indexed by host name.
    
    @ivar connections: Thread-local connection pool.
    :type connections: C{dict} of C{str} to C{httplib.HTTPConnection}
    """
    def __init__(self):
        self.connections = {}
        
pool = Pool()

class TLSConnectionPoolMixin(object):
    """
    A mixin for Transport classes that attempts to use connections from a thread-local-storage 
    pool before creating new connections.
    
    The connections are indexed by host.  This will NOT work if you are mixing HTTP and
    HTTPS connections to the same host!!!
    
    :param host: The host for which connection is being requested.
    :type host: C{str}
    
    :return: An existing or new C{httplib.HTTPConnection}
    :rtype: C{httplib.HTTPConnection}
    """
    
    def request(self, host, handler, body, headers=None, verbose=False):
        """
        Override to add Connection: keep-alive header to request.
        """
        if headers is None:
            headers = {}
        headers['Connection'] = 'keep-alive'
        return super(TLSConnectionPoolMixin, self).request(host, handler, body, headers=headers, verbose=verbose)
        
    def connect(self, host):
        """
        Overrides method to return an existing connection from thread-local pool
        instead of creating a new one.
        """
        if not host in pool.connections:
            self.logger.debug("No connection in pool for %s, creating." % host)
            conn = super(TLSConnectionPoolMixin, self).connect(host)
            pool.connections[host] = conn
        else:
            self.logger.debug("Found EXISTING connection in pool for %s." % host)
            conn = pool.connections[host]
        return conn    

    def handle_connection_error(self, host, x):
        """
        Remove the offending connection from the pool.
        """
        self.logger.info('Deleting bad connection to host %s' % host)
        del pool.connections[host]

rpctools/jsonrpc/transport.py
from __future__ import absolute_import

import sys
import socket
import httplib
import logging

from rpctools.jsonrpc import ssl_wrapper
from rpctools.jsonrpc.exc import ConnectionError, ProtocolError
from rpctools.jsonrpc.pool import TLSConnectionPoolMixin

__license__ = """Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""

class Transport(object):
    """
    Handles an HTTP transaction to a JSON-RPC server.
    
    :ivar user_agent: The user agent to report when connecting to server.
    :type user_agent: C{str}
    
    :ivar timeout: The socket timeout (in seconds) to use on httplib connections. 
                        (defaults to socket._GLOBAL_DEFAULT_TIMEOUT)
    :type timeout: C{int}
    """

    user_agent = "JSON-RPC Client"    
    timeout = socket._GLOBAL_DEFAULT_TIMEOUT

    def __init__(self, timeout=None):
        self.logger = logging.getLogger('{0.__name__}.{0.__module__}'.format(self.__class__))
        if timeout is not None:
            self.timeout = timeout

    def request(self, host, handler, body, headers=None, verbose=False):
        """
        Send a complete request, and parse the response.
        
        :param host: Target host (may include port, e.g. 'example.com:8080')
        :type host: C{str}
        
        :param handler: Target PRC handler (e.g. '/jsonrpc').
        :type handler: C{str}
        
        :param body: Request body. (Assumes it is already correctly formatted/encoded.)
        :type body: C{str}
        
        :param headers: HTTP headers to send with request.
        :type headers: C{dict}
        
        :param verbose: Debugging flag.
        :type verbose: C{bool}
        
        :return: The response to the request.
        :rtype: C{httplib.HTTPResponse}
        
        :raise ProtocolError: If the response status is not 200.
        """
        if headers is None:
            headers = {}
        
        conn = self.connect(host)
        
        if verbose:
            conn.set_debuglevel(1)
        
        header_keys_lower = set(k.lower() for k in headers.keys())  
        
        # Add headers
        headers['User-Agent'] = self.user_agent
        if not 'content-type' in header_keys_lower:
            headers['Content-Type'] = 'application/json'
        headers['Content-Length'] = len(body) if body is not None else 0
        
        try:
            conn.request("POST", handler, body, headers)
            response = conn.getresponse()
            
            if response.status != 200:
                raise ProtocolError(host + handler, response.status, response.reason, headers)
            
            return response
        except (socket.error, httplib.HTTPException) as x:
            self.handle_connection_error(host, x)
            exc_class, exc, tb = sys.exc_info()
            cerror = ConnectionError("Error connecting to host %s: %r" % (host, x))
            raise ConnectionError, cerror, tb
        
    def handle_connection_error(self, host, x):
        """
        Stub method to handle connection errors for specified host.
        
        (This exists to support removing things from pool, etc.)
        
        :param host: The host associated with the error.
        :type host: str
        """
        
    def connect(self, host):
        """
        Connect to specified server.
        
        This method returns a new connection.  (It should be overridden if you would like
        to change that behavior -- e.g. to re-use existing connections.)
        
        :param host: The host (optionally in "host:port" syntax).
        :type host: C{str}
        
        :return: A connection handle.
        :rtype: C{httplib.HTTPConnection}
        """
        return httplib.HTTPConnection(host, timeout=self.timeout)


class SafeTransport(Transport):
    """
    Extends/overrides Transport to use HTTPS connections.
    """
    
    def __init__(self, key_file=None, cert_file=None, ca_certs=None, validate_cert_hostname=True, timeout=None):
        super(SafeTransport, self).__init__(timeout=timeout)
        self.key_file = key_file
        self.cert_file = cert_file
        self.ca_certs = ca_certs
        self.validate_cert_hostname = validate_cert_hostname
        
    def connect(self, host):
        """
        Connect securely (HTTPS) to host.
        """
        return ssl_wrapper.CertValidatingHTTPSConnection(host, key_file=self.key_file, cert_file=self.cert_file,
                   ca_certs=self.ca_certs, validate_cert_hostname=self.validate_cert_hostname)


class TLSConnectionPoolTransport(TLSConnectionPoolMixin, Transport):
    pass

class TLSConnectionPoolSafeTransport(TLSConnectionPoolMixin, SafeTransport):
    pass


rpctools/jsonrpc/client.py
"""
Client libraries for connecting over JSON-RPC, protocol version 1.0.

These are based on the libraries in xmlrpclib and on code from: U{http://code.activestate.com/recipes/552751/};
however, they also add support for using keep-alive connections.
"""
from __future__ import absolute_import

import logging
import urllib
import json
import base64
import string

from Cookie import SimpleCookie

from rpctools.jsonrpc.transport import Transport, SafeTransport, TLSConnectionPoolSafeTransport, TLSConnectionPoolTransport
from rpctools.jsonrpc.exc import JsonRpcError, ResponseError, Fault

__license__ = """Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
 
  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""

# -----------------------------------------------------------------------------
# "INTERNAL" CLASSES
# -----------------------------------------------------------------------------

class _Method(object):
    """
    Some magic to bind an JSON-RPC method to an RPC server.
    
    Supports "nested" methods (e.g. examples.getStateName).
    """
    def __init__(self, send, name):
        """
        Initialize this proxy callable with send function and method name.
        
        :param send: A callable that will perform the actual request.
        :type send: C{callable}
        
        :param name: The method name.
        :type name: C{str}
        """
        self._send = send
        self._name = name
        
    def __getattr__(self, name):
        """
        Dereferences this callable to support "nested" methods.
        
        This supports having a "module" on server that contains
        methods of its own.  (E.g. "examples.getStateName")
        
        :rtype: L{_Method}
        """
        return _Method(self._send, "%s.%s" % (self._name, name))
    
    def __call__(self, *args, **kwargs):
        """
        Performs the actual request, using the configured request sender. 
        """
        if args and kwargs:
            raise JsonRpcError('JSON-RPC 2.0 spec does not allow both positional and keyword arguments.')
        return self._send(self._name, args if args else kwargs)
    
    def __repr__(self):
        return '<%s name=%s>' % (self.__class__.__name__, self._name)

# -----------------------------------------------------------------------------
# PUBLIC CLASSES
# -----------------------------------------------------------------------------

class ServerProxy(object):
    """ 
    The ServerProxy provides a proxy to the remote JSON-RPC service methods and performs the
    request encoding and decoding. 
    
    This class uses instance variables to save state and is NOT THREAD-SAFE.
    
    :ivar type: The scheme we're using for connection: 'http' or 'https'
    :type type: C{str}
     
    :ivar host: The host we're connecting to (may include port, e.g. "foobar.com:8080")
    :type host: C{str}
    
    :ivar transport: A L{Transport} instance to use.
    :type transport: L{Transport}
    
    :ivar id: The JSON-RPC 'id' field (incrementing integer).
    :type id: C{int}
    
    :ivar extra_headers: Any HTTP request headers that should be sent with every request.
    :type extra_headers: C{dict}
    
    :ivar method_class: The proxy class for the remote methods (default is L{_Method}).
    :type method_class: C{type}
    """
    
    method_class = _Method 
    
    def __init__(self, uri, key_file=None, cert_file=None, ca_certs=None, validate_cert_hostname=True,
                 extra_headers=None, timeout=None, pool_connections=False):
        """
        :param uri: The endpoint JSON-RPC server URL.
        :param key_file: Secret key to use for ssl connection.
        :param cert_file: Cert to send to server for ssl connection.
        :param ca_certs: File containing concatenated list of certs to validate server cert against.
        :param extra_headers: Any additional headers to include with all requests.
        :param pool_connections: Whether to use a thread-local connection pool for connections.
        """
        self.logger = logging.getLogger('{0.__module__}.{0.__name__}'.format(self.__class__))
        if extra_headers is None:
            extra_headers = {}
            
        self.type, remainder = urllib.splittype(uri)
        if self.type not in ("http", "https"):
            raise JsonRpcError("unsupported JSON-RPC uri: %s" % uri)
        
        (host_part, self.handler) = urllib.splithost(remainder)
        
        # If we have auth info in our host part, remove it and add it to extra headers
        # (This is based on impl in xmlrpclib)
        (auth, self.host) = urllib.splituser(host_part)
        if auth:
            auth = base64.encodestring(urllib.unquote(auth))
            auth = string.join(string.split(auth), "") # get rid of whitespace
            extra_headers.update({"Authorization": "Basic " + auth})
        
        self.key_file = key_file
        self.cert_file = cert_file
        self.ca_certs = ca_certs
        self.validate_cert_hostname = validate_cert_hostname
        
        # TODO: This could probably be a little cleaner :)
        if pool_connections:
            if self.type == "https":
                self.transport = TLSConnectionPoolSafeTransport(timeout=timeout)
            else:
                self.transport = TLSConnectionPoolTransport(timeout=timeout)
        else:
            if self.type == "https":
                self.transport = SafeTransport(key_file=self.key_file, cert_file=self.cert_file,
                    ca_certs=self.ca_certs, validate_cert_hostname=self.validate_cert_hostname)
            else:
                self.transport = Transport(timeout=timeout)
        
        self.extra_headers = extra_headers
        self.id = 0 # Initialize our request ID (gets incremented for every request)
        
    def _request(self, methodname, params):
        """
        Overriden __request method which introduced the request_cookie parameter
        that allows cookie headers to propagated through JSON-RPC requests.
        
        :param methodname: Name of method to be called.
        :type methodname: C{str}
        
        :param params: Parameters list to send to method.
        :type params: C{list}
        
        :return: The decoded result.
        
        :raise ResponseError: If the response cannot be parsed or is not proper JSON-RPC 1.0 response format.
        :raise ProtocolError: Re-raises exception if non-200 response received.
        :raise Fault: If the response is an error message from remote application.
        """
        self.id += 1 # Increment our "unique" identifier for every request.
        
        data = dict(id=self.id, method=methodname, params=params)
        
        headers = self.extra_headers
        
        self._prepare_request(data, headers)
        
        body = json.dumps(data)
        
        response = self.transport.request(self.host, self.handler, body, headers=headers)
        
        self._handle_response(response)
        
        data = response.read()
        
        try:
            decoded = json.loads(data)
        except Exception as x:
            raise ResponseError("Unable to parse response data as JSON: %s" % x)
        
        # This special casing for non-compliant systems like DD that sometimes
        # just return NULL from actions and think they're communicating w/ valid
        # JSON-RPC.
        if decoded is None:
            return None
        
        if not (('result' in decoded) or ('error' in decoded)):
            # Include the decoded result (or part of it) in the error we raise
            r = repr(decoded)
            if len(r) > 256: # a hard-coded value to keep the exception message to a sane length
                r = r[0:255] + '...'
            raise ResponseError('Malformed JSON-RPC response to %s: %s' % (methodname, r))
        
        if 'error' in decoded and decoded['error']:
            raise Fault(decoded['error']['code'], decoded['error']['message'])
        
        if not 'result' in decoded:
            raise ResponseError('Malformed JSON-RPC response: %r' % decoded)
         
        return decoded['result']

    def _prepare_request(self, data, headers):
        """
        An extension point hook for preparing the request data before it is encoded
        and sent to server.
        
        This method may modify the body (C{dict}) and headers (C{dict}).  Note
        that some headers (e.g. Content-Length) may be added by the underlying
        libraries (e.g. httplib).
        
        :param data: The request data (C{dict}) that will be sent to server.
        :type data: C{dict}
        
        :param headers: Headers that will be sent with the request.  This includes
                        any headers from the extra_headers instance var.
        :type headers: C{dict}
        """
        pass
    
    def _handle_response(self, response):
        """
        An extension point hook for processing the raw response objects from the server.
        
        :param response: The HTTP response object.
        :type response: C{httplib.HTTPResponse}
        """
        pass
    
    def __getattr__(self, name):
        """
        Does the proxy magic: returns a proxy callable that will perform request (when called).
        
        :return: The callable method object which will perform the request to JSON-RPC server.
        :rtype: L{_Method}
        """
        return self.method_class(self._request, name)
    
    def __repr__(self):
        return ("<%s for %s%s>" % (self.__class__.__name__, self.host, self.handler))


class RawServerProxy(ServerProxy):
    """ 
    An "extension" of L{ServerProxy} that does not attempt to parse the response as a JSON-RPC
    message, but simply returns the response data and file-like object. 
    """
    
    def _request(self, methodname, params):
        """
        Overriden __request method which stops short of performing any parsing of the response.
        
        :param methodname: Name of method to be called.
        :type methodname: C{str}
        
        :param params: Parameters list to send to method.
        :type params: C{list}
        
        :return: The C{httplib.HTTPResponse} object (which contains headers, body, etc.)
        :rtype: C{httplib.HTTPResponse}
        :raise ProtocolError: Re-raises exception if non-200 response received.
        """
        self.id += 1 # Increment our "unique" identifier for every request.
        
        data = dict(id=self.id, method=methodname, params=params)
        
        headers = self.extra_headers
        
        self._prepare_request(data, headers)
        
        body = json.dumps(data)
        
        response = self.transport.request(self.host, self.handler, body, headers=headers)
        
        self._handle_response(response)
        
        return response

class CookieKeeperMixin(object):
    """
    A L{ServerProxy} that supports receiving and setting cookies.
    
    IMPORTANT: If this class is being used to support cookie-based sessions,
    it should be noted that it will only support a single session.  I.e. a
    single instance of this class should only pool connections for a single
    session!
    
    THIS CLASS IS NOT THREAD SAFE.
     
    :ivar response_cookies: A dict of all cookies sent from server (Set-Cookie headers), indexed by cookie name.
    :type response_cookies: C{dict} of C{str} to C{Cookie.SimpleCookie}
    
    :ivar request_cookies: A dict of all cookies that will be sent to server, indexed by cookie name.
    :type request_cookies: C{dict} of C{str} to C{Cookie.SimpleCookie}
    
    :ivar auto_add_cookies: Whether to automatically append any received response cookies to requests.
    :type auto_add_cookies: C{bool}
    """
    auto_add_cookies = False
    
    def __init__(self, *args, **kwargs):
        """
        :keyword request_cookies: A dict of request cookies that should be included.
        :type request_cookies: C{dict} 
        """
        self.response_cookies = {}
        self.request_cookies = {}
        if 'request_cookies' in kwargs:
            self.request_cookies = kwargs.pop('request_cookies')
        super(CookieKeeperMixin, self).__init__(*args, **kwargs)    
    
    def add_cookie(self, cookie):
        """
        Adds a *copy* of specified cookie to the request.
        
        Note that a single C{Cookie.SimpleCookie} instance can actually hold multiple cookie 
        names/values.  If this is the case, the cookie will be broken into individual 
        cookies for each "morsel".
        
        :param cookie: The cookie to add to the request.
        :type cookie: C{Cookie.SimpleCookie}
        """
        for (name, morsel) in cookie.items():
            self.request_cookies[name] = SimpleCookie(morsel.output())
        
    def _prepare_request(self, data, headers):
        """
        Extends base implementation to add any saved cookies to the request headers.
        
        :param data: The request data (C{dict}) that will be sent to server.
        :type data: C{dict}
        
        :param headers: Headers that will be sent with the request.  This includes
                        any headers from the extra_headers instance var.
        :type headers: C{dict}
        """
        super(CookieKeeperMixin, self)._prepare_request(data, headers)
        
        req_cookies = {}
        if self.auto_add_cookies and self.response_cookies:
            req_cookies.update(self.response_cookies)
        
        if self.request_cookies:
            req_cookies.update(self.request_cookies)
        
        for cookie in req_cookies.values():
            for name, morsel in cookie.items():
                headers["Cookie"] = "%s=%s" % (name, morsel.value)
            
    def _handle_response(self, response):
        """
        Extends base implementation to extract any cookies from the response.
        
        :param response: The HTTP response object.
        :type response: C{httplib.HTTPResponse}
        """
        super(CookieKeeperMixin, self)._handle_response(response)
        if len(response.msg.getheaders("Set-Cookie")):
            # Build a dict of cookies; we only want to keep the *last* cookie 
            # set with a specific name.
            cookies = {}
            for hdr in response.msg.getheaders("Set-Cookie"):
                cookie = SimpleCookie(hdr)
                # XXX: This is a bit ugly looking.  Basically we are only going to
                # get one cookie at a time from this header; however, the SimpleCookie
                
                # object is designed to support multiple cookie names/values.
                key = cookie.keys()[0]
                cookies[key] = cookie
                
            self.response_cookies = cookies
            
            
class CookieAwareServerProxy(CookieKeeperMixin, ServerProxy):
    pass

class CookieAwareRawServerProxy(CookieKeeperMixin, RawServerProxy):
    pass
